Move cloud-controller-manager to staging k8s.io/cloud-provider

cici37 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
This PR is to move code under cmd/cloud-controller-manager/app to k8s.io/cloud-provider.
Will close the issue https://github.com/kubernetes/cloud-provider/issues/29

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/cloud-provider/issues/29

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

cici37 committed:
Move CCM to staging k8s.io/cloud-provider

cici37 commented:
PR merge currently is blocked by https://github.com/kubernetes/kubernetes/pull/91474. 
Will remove WIP after the above PR merged.

cici37 commented:
/sig cloud-provider

cici37 commented:
/test pull-kubernetes-bazel-test

cici37 commented:
cc @cheftako @andrewsykim

fedebongio commented:
/assign @cheftako 
/triage accepted

cici37 commented:
Hello @cheftako @andrewsykim , This PR is ready to be reviewed :)

cici37 commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

cheftako commented:
/milestone v1.20

cici37 commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

cici37 commented:
Hi @cheftako @andrewsykim , would you please review this PR when have time? Thank you 🙏

cheftako commented:
/lgtm

sttts commented:
/approve

dims commented:
/assign @liggitt

cheftako commented:
/approve

liggitt commented on staging/src/k8s.io/cloud-provider/app/config/config.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package app
+package config
```
this was weird before... thanks for fixing this

liggitt commented on cmd/cloud-controller-manager/.import-restrictions:
```diff
@@ -1,84 +0,0 @@
-rules:
```
rather than eliminating this file, can we change it to this to ensure no new k8s.io/kubernetes deps get added?

```
  - selectorRegexp: k8s[.]io/kubernetes
    allowedPrefixes:
      - k8s.io/kubernetes/cmd/cloud-controller-manager
```

cheftako replied:
+1

liggitt reviewed (commented):
one comment on the import restrictions, and needs a rebase

andrewsykim commented on test/integration/serving/serving_test.go:
```diff
@@ -32,7 +32,7 @@ import (
 	"k8s.io/cloud-provider"
 	"k8s.io/cloud-provider/fake"
-	cloudctrlmgrtesting "k8s.io/kubernetes/cmd/cloud-controller-manager/app/testing"
+	cloudctrlmgrtesting "k8s.io/cloud-provider/testing"
```
Would it make sense to keep this in under `app` based on it's previous location (i.e. `k8s.io/cloud-provider/app/testing`)?

andrewsykim commented on staging/src/k8s.io/cloud-provider/config/scheme/scheme.go:
```diff
@@ -20,8 +20,8 @@ import (
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	"k8s.io/cloud-provider/app/apis/config"
-	"k8s.io/cloud-provider/app/apis/config/v1alpha1"
+	"k8s.io/cloud-provider/config"
```
This package should probably be called "install" instead?

andrewsykim replied:
Was referring to `k8s.io/cloud-provider/config/scheme`

cici37 replied:
Was just going to ask^^. Thanks for the further explain :)

cici37 replied:
Then I might also change the func name to install instead of `AddToSchema`

andrewsykim reviewed (commented):
All my comments are unblocking, thanks @cici37 !

cici37 commented:
Updated the pr since last review to:
1. Added back .import-restrictions to cmd/cloud-controller-manager to ensure no new k8s.io/kubernetes deps get added.
2. Moved `k8s.io/cloud-provider/testing` to `k8s.io/cloud-provider/app/testing`
3. Changed package name from `k8s.io/cloud-provider/config/scheme` to `k8s.io/cloud-provider/config/install` and also changed the func name inside to `install`.

@liggitt Would you please have a look again when you have time? Thank you 🙏

cici37 commented:
/test pull-kubernetes-node-e2e

liggitt commented:
/approve

@cheftako has lgtm

cheftako commented:
/lgtm

cici37 commented:
/retest

cici37 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

cici37 commented:
/retest
